imap_processing/tests/ultra/unit/test_de.py

```diff
@@ -0,0 +1,119 @@
+
+    # Coincidence positions
+    assert np.allclose(dataset["x_coin"], df_filt["Xc"].astype("float"))
+
+    # Time of flight
+    assert np.allclose(dataset["tof_start_stop"], df_filt["TOF"].astype("float"))
+    assert np.allclose(dataset["tof_stop_coin"], df_filt["eTOF"].astype("float"))
+    assert np.allclose(dataset["tof_corrected"], df_filt["cTOF"].astype("float"))
+
+    # Distances and path lengths
+    assert np.allclose(dataset["front_back_distance"], df_filt["d"].astype("float"))
+    assert np.allclose(dataset["path_length"], df_filt["r"].astype("float"))
+
+    # Coincidence, start, and event types
+    assert np.allclose(dataset["coincidence_type"], df_filt["CoinType"].astype("float"))
+    assert np.allclose(dataset["start_type"], df_filt["StartType"].astype("float"))
+    assert np.allclose(dataset["event_type"], df_filt["StopType"].astype("float"))
+
+    # Energies and species
+    assert np.allclose(dataset["energy"], df_filt["Energy"].astype("float"))
+    assert np.allclose(
+        dataset["species"], np.full(len(de_dataset["epoch"]), np.nan, dtype=np.uint8)
+    )
+
+    # Velocities in various frames
+    test_tof = dataset["tof_start_stop"]
+    assert np.allclose(
+        dataset["vx_ultra"][test_tof > 0],
+        -df_filt["vhatX"].astype("float").values[test_tof > 0],
+        rtol=1e-2,
+    )
+    assert np.allclose(
+        dataset["vy_ultra"][test_tof > 0],
+        -df_filt["vhatY"].astype("float").values[test_tof > 0],
+        rtol=1e-2,
+    )
+    assert np.allclose(
+        dataset["vz_ultra"][test_tof > 0],
+        -df_filt["vhatZ"].astype("float").values[test_tof > 0],
+        rtol=1e-2,
+    )
+
+    # Spacecraft and heliocentric velocities (nan comparisons)
+    assert np.allclose(
+        dataset["vx_sc"].data, np.full(len(de_dataset["epoch"]), np.nan), equal_nan=True
+    )
+    assert np.allclose(
+        dataset["vy_sc"].data, np.full(len(de_dataset["epoch"]), np.nan), equal_nan=True
+    )
+    assert np.allclose(
+        dataset["vz_sc"].data, np.full(len(de_dataset["epoch"]), np.nan), equal_nan=True
+    )
+
+    assert np.allclose(
+        dataset["vx_dps_sc"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
+    assert np.allclose(
+        dataset["vy_dps_sc"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
+    assert np.allclose(
+        dataset["vz_dps_sc"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
+
+    assert np.allclose(
+        dataset["vx_dps_helio"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
+    assert np.allclose(
+        dataset["vy_dps_helio"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
+    assert np.allclose(
+        dataset["vz_dps_helio"].data,
+        np.full(len(de_dataset["epoch"]), np.nan),
+        equal_nan=True,
+    )
```

Comment:
Why are we testing these against all NaN arrays? Do we want them to populate with NaNs, or is it that we don't have valid test data, but still want to ensure the arrays are created with the proper shape?

If the latter, could we either use fake data which would result in round numbers for testing or just test the shape of these arrays?

Reply:
These are all calculated here sc_velocity, sc_dps_velocity, helio_velocity = get_annotated_particle_velocity.

And since we don't have SPICE kernels for the timerange of the test data we can't test them. I can add a shape test instead.

Reply:
Thanks for clarifying - sounds good!